File: webapp/packages/grounding/src/oklab.ts
/**
 * OKLab space implementation for TypeScript.
 */

import type { OKLab } from '@oklab/core';
import type { Space, GroundRegion } from './space';

// OKLab space implementation
export class OKLabSpace implements Space<OKLab> {
  distance(a: OKLab, b: OKLab): number {
    const dl = a.L - b.L;
    const da = a.a - b.a;
    const db = a.b - b.b;
    return Math.sqrt(dl * dl + da * da + db * db);
  }

  mix(points: OKLab[], weights: number[]): OKLab {
    if (points.length === 0) {
      throw new Error('Cannot mix empty list of points');
    }
    if (points.length !== weights.length) {
      throw new Error('Points and weights must have same length');
    }

    // Normalize weights
    const totalWeight = weights.reduce((sum, w) => sum + w, 0);
    const normalizedWeights = totalWeight === 0
      ? weights.map(() => 1 / points.length)
      : weights.map(w => w / totalWeight);

    // Mix in OKLab space
    const mixedL = points.reduce((sum, p, i) => sum + p.L * normalizedWeights[i], 0);
    const mixedA = points.reduce((sum, p, i) => sum + p.a * normalizedWeights[i], 0);
    const mixedB = points.reduce((sum, p, i) => sum + p.b * normalizedWeights[i], 0);

    return { L: mixedL, a: mixedA, b: mixedB };
  }

  validate(point: OKLab): boolean {
    // OKLab has L in [0,1], a,b typically in [-1,1] but can be wider
    return point.L >= 0 && point.L <= 1 &&
           point.a >= -1 && point.a <= 1 &&
           point.b >= -1 && point.b <= 1;
  }
}

// Spherical region around a center point
export class SphericalRegion implements GroundRegion<OKLab> {
  constructor(
    public center: OKLab,
    public radius: number,
    private space: OKLabSpace
  ) {}

  contains(point: OKLab): boolean {
    return this.space.distance(this.center, point) <= this.radius;
  }
}

// Utility for creating color regions
export function createColorRegion(colorName: string, space: OKLabSpace): SphericalRegion {
  // Simplified color mapping - in practice would use OKLab values for actual colors
  const colorMap: Record<string, OKLab> = {
    "red": { L: 0.5, a: 0.3, b: 0.2 },
    "blue": { L: 0.4, a: -0.1, b: -0.3 },
    "green": { L: 0.6, a: -0.2, b: 0.2 },
    "yellow": { L: 0.8, a: 0.1, b: 0.2 },
    "purple": { L: 0.4, a: 0.1, b: -0.2 }
  };

  const center = colorMap[colorName.toLowerCase()] || { L: 0.5, a: 0, b: 0 };
  return new SphericalRegion(center, 0.1, space);
}

File: oklab_grounding/oklab.py
"""
OKLab space implementation for the grounding framework.
"""

from typing import List, Tuple
import math
from .space import GroundSpace, Symbol
from .numeric import quantize, qtuple, clamp_angle_pi

# OKLab color type
class OKLab:
    def __init__(self, L: float, a: float, b: float):
        self.L = L
        self.a = a
        self.b = b

    def __repr__(self):
        return f"OKLab(L={self.L}, a={self.a}, b={self.b})"

    def __eq__(self, other):
        if not isinstance(other, OKLab):
            return False
        return (self.L == other.L and
                self.a == other.a and
                self.b == other.b)

    def __iter__(self):
        yield self.L
        yield self.a
        yield self.b

class OKLabSpace(GroundSpace[OKLab]):
    """
    OKLab color space implementation of GroundSpace.

    Provides perceptual distance, color mixing, and validation for OKLab colors.
    """

    def distance(self, a: OKLab, b: OKLab) -> float:
        """Perceptual distance in OKLab space."""
        dl = a.L - b.L
        da = a.a - b.a
        db = a.b - b.b
        return math.sqrt(dl*dl + da*da + db*db)

    def mix(self, points: List[OKLab], weights: List[float]) -> OKLab:
        """Convex combination of OKLab colors."""
        if not points:
            raise ValueError("Cannot mix empty list of points")
        if len(points) != len(weights):
            raise ValueError("Points and weights must have same length")

        # Normalize weights
        total_weight = sum(weights)
        if total_weight == 0:
            # Equal weights if all zero
            normalized_weights = [1.0 / len(points)] * len(points)
        else:
            normalized_weights = [w / total_weight for w in weights]

        # Mix in OKLab space
        mixed_L = sum(p.L * w for p, w in zip(points, normalized_weights))
        mixed_a = sum(p.a * w for p, w in zip(points, normalized_weights))
        mixed_b = sum(p.b * w for p, w in zip(points, normalized_weights))

        return OKLab(mixed_L, mixed_a, mixed_b)

    def validate(self, point: OKLab) -> bool:
        """Check if OKLab color is valid (within reasonable bounds)."""
        # OKLab has L in [0,1], a,b typically in [-0.5, 0.5] but can be wider
        return (0 <= point.L <= 1 and
                -1 <= point.a <= 1 and
                -1 <= point.b <= 1)

# Predefined color regions for common symbols
class SphericalRegion:
    """Simple spherical region around a center point."""

    def __init__(self, center: OKLab, radius: float, space: OKLabSpace):
        self.center = center
        self.radius = radius
        self.space = space

    def contains(self, point: OKLab) -> bool:
        return self.space.distance(self.center, point) <= self.radius

# Utility functions for creating common color regions
def create_color_region(color_name: str, space: OKLabSpace) -> 'SphericalRegion':
    """Create a region for a named color."""
    # Simplified color mapping - in practice would use OKLab values for actual colors
    color_map = {
        "red": OKLab(0.5, 0.3, 0.2),
        "blue": OKLab(0.4, -0.1, -0.3),
        "green": OKLab(0.6, -0.2, 0.2),
        "yellow": OKLab(0.8, 0.1, 0.2),
        "purple": OKLab(0.4, 0.1, -0.2)
    }

    center = color_map.get(color_name.lower(), OKLab(0.5, 0, 0))
    return SphericalRegion(center, 0.1, space)  # Small radius for distinct regions